crates/red_knot_python_semantic/src/types.rs

```diff
@@ -482,26 +495,39 @@ impl<'db> Type<'db> {
     ///
     /// Returns `None` if `self` is not a callable type.
     #[must_use]
-    pub fn call(&self, db: &'db dyn Db) -> Option<Type<'db>> {
+    fn call(self, db: &'db dyn Db, arg_types: &[Type<'db>]) -> CallOutcome<'db> {
         match self {
-            Type::Function(function_type) => Some(function_type.return_type(db)),
+            // TODO validate typed call arguments vs callable signature
+            Type::Function(function_type) => CallOutcome::callable(function_type.return_type(db)),
+            Type::RevealTypeFunction(function_type) => CallOutcome::RevealType {
+                return_ty: function_type.return_type(db),
+                revealed_ty: *arg_types.first().unwrap_or(&Type::Unknown),
```

Comment:
Should this be `None` instead for better error messages? It's a difference whether a type is inferred as `Unknown` or the argument was missing.

Reply:
Possibly. But then you end up with two diagnostics where one is the error and the other is the reveal type diagnostic which doesn't seem ideal to me. But I can see that unifying the diagnostic is probably more work.

Reply:
Good point that it may not make sense to even have a revealed-type diagnostic if the call to `reveal_type` is incorrect. But when we add call-signature checking (which can be shared in a common function used across both the `Function` and `RevealTypeFunction` arms, because it just needs the call arguments and the `FunctionType`), the `RevealTypeFunction` arm isn't required to return a `CallOutcome::RevealType` at all. If it fails call signature checking, it can return a different outcome type, that won't issue a revealed-type diagnostic at all.

So in that case you can consider this fallback just a temporary band-aid until we have call signature checking (which I already added a TODO for above.)